enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,494 @@
+
+	Spec   BootImageHistorySpec   `json:"spec,omitempty"`
+	Status BootImageHistoryStatus `json:"status,omitempty"`
+}
+
+// BootImageHistorySpec defines the desired state of BootImageHistory
+type BootImageHistorySpec struct {
+}
+
+// BootImageHistoryStatus defines the observed state of BootImageHistory
+type BootImageHistoryStatus struct {
+	// machineResourceReference contains identifying information of the machine management resource being tracked.
+	// +kubebuilder:validation:Required
+	// +required
+	MachineResourceReference MachineResourceReference `json:"machineResourceReference"`
+	// details is a list of boot image history entries of the machine resource.
+	// +optional
+	Details []BootImageHistoryDetail `json:"details,omitempty"`
+}
+
+type MachineResourceReference struct {
+	// name is the machine management resource's name
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+	// kind is the machine management resource's kind
+	// +kubebuilder:validation:Required
+	// +required
+	Kind string `json:"kind"`
+	// apiGroup is name of the APIGroup that the machine management resource belongs to. This is for disambiguating
+	// between Cluster API and Machine API backed resources.
+	// +kubebuilder:validation:Required
+	// +required
+	APIGroup string `json:"apiGroup"`
+}
+
+// BootImageHistoryDetail is the struct for each element in the Details array
+type BootImageHistoryDetail struct {
+	// updateTime records the timestamp at which the update took place.
+	// +required
+	UpdateTime metav1.Time `json:"updatedTime"`
+	// bootImageRef records the new boot image reference to which the update took place.
+	// +required
+	BootImageRef string `json:"bootImageRef"`
+}
+
+// BootImageHistoryList contains a list of BootImageHistory
+type BootImageHistoryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []BootImageHistory `json:"items"`
+}
+
+```
+There will be one instance of this per machine management resource(which can be a MachineSet[MAPI or CAPI], MachineDeployment...etc). It will be named the same as the resource being tracked. The MSBIC is responsible for creating and updating this CR when a boot image update takes place. This CR will exist in the same namespace as the resource.
+
+YAML Example for a MAPI backed machineset scenario:
+```
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: BootImageHistory
+metadata:
+  name: djoshy10-2tcqv-worker-a
+spec: {}
+status:
+  machineResourceReference:
+    name: djoshy10-2tcqv-worker-a
+    kind: MachineSet
+    apiGroup: cluster.x-k8s.io/v1alpha3
+  details:
+    - updateTime: "2023-12-14T12:00:00Z"
+      bootImageRef: "projects/rhcos-cloud/global/images/rhcos-414-92-202308032115-0-gcp-x86-64"
+    - updateTime: "2023-12-14T14:30:00Z"
+      bootImageRef: "projects/rhcos-cloud/global/images/rhcos-415-92-202311241643-0-gcp-x86-64"
+
+```
+
+YAML Example for a CAPI backed machineset scenario:
```

Comment:
Heads up, this description is inverted. This is a MAPI backed machine, whereas the above YAML is the CAPI one.

Reply:
Thanks! Will fix on next pass (: